rustls/src/crypto/mod.rs

```diff
@@ -45,19 +45,137 @@ pub use crate::rand::GetRandomFailed;
+/// `SigningKey`, etc.) come from the same crate.  It is allowed and expected that uninteresting
+/// elements would be delegated back to one of the default providers (statically) or a parent
+/// provider (dynamically).
+///
+/// For example, if we want to make a provider that just overrides key loading in the config builder
+/// API ([`crate::ConfigBuilder::with_single_cert`] etc.), it might look like this:
+///
+/// ```
+/// # #[cfg(feature = "ring")] {
+/// # use std::sync::Arc;
+/// # mod fictious_hsm_api { pub fn load_private_key(key_der: pki_types::PrivateKeyDer<'static>) -> ! { unreachable!(); } }
+/// use rustls::crypto::ring::RING;
+///
+/// #[derive(Debug)]
+/// struct HsmKeyLoader;
+///
+/// impl rustls::crypto::CryptoProvider for HsmKeyLoader {
+///     fn fill_random(&self, buf: &mut [u8]) -> Result<(), rustls::crypto::GetRandomFailed> {
+///         RING.fill_random(buf)
+///     }
+///
+///     fn default_cipher_suites(&self) -> &'static [rustls::SupportedCipherSuite] {
+///         RING.default_cipher_suites()
+///     }
+///
+///     fn default_kx_groups(&self) -> &'static [&'static dyn rustls::crypto::SupportedKxGroup] {
+///         RING.default_kx_groups()
+///     }
+///
+///     fn signature_verification_algorithms(&self) -> rustls::WebPkiSupportedAlgorithms {
+///         RING.signature_verification_algorithms()
+///     }
+///
+///     fn load_private_key(&self, key_der: pki_types::PrivateKeyDer<'static>) -> Result<Arc<dyn rustls::sign::SigningKey>, rustls::Error> {
+///         fictious_hsm_api::load_private_key(key_der)
+///     }
+/// }
+/// # }
+/// ```
+///
+/// ## References to the individual elements
+/// The elements are documented separately:
+///
+/// - **Random** - see [`crate::crypto::CryptoProvider::fill_random()`].
+/// - **Cipher suites** - see [`crate::SupportedCipherSuite`], [`crate::Tls12CipherSuite`], and
+///   [`crate::Tls13CipherSuite`].
+/// - **Key exchange groups** - see [`crate::crypto::SupportedKxGroup`].
+/// - **Signature verification algorithms** - see [`crate::WebPkiSupportedAlgorithms`].
+/// - **Authentication key loading** - see [`crate::crypto::CryptoProvider::load_private_key()`] and
+///   [`crate::sign::SigningKey`].
+///
+/// # Example code
+/// See [provider-example/] for a full client and server example that uses
+/// cryptography from the [rust-crypto] and [dalek-cryptography] projects.
+///
+/// ```shell
+/// $ cargo run --example client | head -3
+/// Current ciphersuite: TLS13_CHACHA20_POLY1305_SHA256
+/// HTTP/1.1 200 OK
+/// Content-Type: text/html; charset=utf-8
+/// Content-Length: 19899
+/// ```
+///
+/// [provider-example/]: https://github.com/rustls/rustls/tree/main/provider-example/
+/// [rust-crypto]: https://github.com/rustcrypto
+/// [dalek-cryptography]: https://github.com/dalek-cryptography
 pub trait CryptoProvider: Send + Sync + Debug + 'static {
     /// Fill the given buffer with random bytes.
+    ///
+    /// The bytes must be cryptographically strong.
```

Comment:
Bytes by themselves are never cryptographically strong, right?

Reply:
Agreed I think this might be clearer if worded to describe that the bytes should be sourced from a CSRNG

Reply:
reworded